BitFaster.Caching/Lfu/TimerWheel.cs

```diff
@@ -0,0 +1,273 @@
+            // so that an overflow only occurs due to using an arbitrary origin time (System.nanoTime()).
+            if ((previousTime < 0) && (currentTime > Duration.Zero))
+            {
+                previousTime += long.MaxValue;
+                currentTime += new Duration(long.MaxValue);
+            }
+
+            try
+            {
+                for (int i = 0; i < Shift.Length; i++)
+                {
+                    long previousTicks = (long)(((ulong)previousTime) >> Shift[i]);
+                    long currentTicks = (long)(((ulong)currentTime.raw) >> Shift[i]);
+                    long delta = (currentTicks - previousTicks);
+                    
+                    if (delta <= 0L)
+                    {
+                        break;
+                    }
+                    
+                    Expire(ref cache, i, previousTicks, delta);
+                }
+            }
+            catch (Exception)
+            {
+                time = previousTime;
+                throw;
+            }
+        }
+
+        // Expires entries or reschedules into the proper bucket if still active.
+        private void Expire<N, P>(ref ConcurrentLfuCore<K, V, N, P> cache, int index, long previousTicks, long delta)
+            where N : LfuNode<K, V>
+            where P : struct, INodePolicy<K, V, N>
+        {
+            TimeOrderNode<K, V>[] timerWheel = wheel[index];
+            int mask = timerWheel.Length - 1;
+
+            // We assume that the delta does not overflow an integer and cause negative steps. This can
+            // occur only if the advancement exceeds 2^61 nanoseconds (73 years).
+            int steps = Math.Min(1 + (int)delta, timerWheel.Length);
+            int start = (int)(previousTicks & mask);
+            int end = start + steps;
+
+            for (int i = start; i < end; i++)
+            {
+                TimeOrderNode<K, V> sentinel = timerWheel[i & mask];
+                TimeOrderNode<K, V> prev = sentinel.GetPreviousInTimeOrder();
+                TimeOrderNode<K, V> node = sentinel.GetNextInTimeOrder();
+                sentinel.SetPreviousInTimeOrder(sentinel);
+                sentinel.SetNextInTimeOrder(sentinel);
+
+                while (node != sentinel)
+                {
+                    TimeOrderNode<K, V> next = node.GetNextInTimeOrder();
+                    node.SetPreviousInTimeOrder(null);
+                    node.SetNextInTimeOrder(null);
+
+                    try
+                    {
+                        // TODO: Caffeine passes the time into evict here, and can resurrect
+                        // https://github.com/ben-manes/caffeine/blob/73d5011f9db373fc20a6e12d1f194f0d7a967d69/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L1023
+                        if ((node.GetTimestamp() - time) < 0)
+                        {
+                            cache.Evict(node);
+                        }

```

Comment:
fyi, prior to the loop the linked list is detached so I rescheduled the node if not expired. Presumably that could be due to a concurrent update so the reordering had not happened yet or similar scenarios. If you don't reschedule then detaching is probably unnecessary. If you do then it avoids an accidental loop, e.g. the overflow bucket (1+ weeks) might scan over items that expire far into the distant future so it would reschedule back into it. Or maybe if it reschedules at the head instead of tail that would resolves it. Just something to review and consider how you prefer handling these scenarios.

Reply:
Thanks! You probably just saved me several hours of debugging - so far I had only tested superficially by replicating some of your unit tests and single stepping in the debugger to understand how it works. Now you pointed this out, it seems that I must call schedule here, otherwise nodes would always stay in the same bucket and never cascade down through the wheels. I had distracted myself trying to understand under what conditions a resurrection could occur and missed it.

I'm not sure if this is very useful to you at this point, but I don't think you have a unit test that would detect my mistake (at least in TimerWheelTest.java), since the only test you have calling `advance` more than once is `advance_backwards`.

I think I have mapped out how to fit everything together, but I need to do a lot more testing. You made a great point in that all time-based policies can be implemented via TimerWheel, so I will start out with that and add the more specialist after access policy later as an optimization.

Reply:
good point! I'll look into adding an explicit unit test for this. I hope that it is implicitly covered indirectly by other tests due to the [data integrity checks](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/test/java/com/github/benmanes/caffeine/cache/LocalCacheSubject.java#L191) that are performed after a cache test passes.

Reply:
I suspected you had it covered - that's effectively what I would like to do next, I will make both a higher-level functional test suite and something that is more like concurrent/stress with constant expiry and check integrity - so good to have that link for inspiration.

Reply:
I ported your test case, works great! 😄

Reply:
> I ported your test case, works great! 😄

It was more or less a copy paste of your other test. It wasn't correct for me - since I am not using nano seconds (Duration can use different time sources with different units to get better perf on different platforms), I had to adjust it a bit to get nodes into each wheel.

I have now added a way to assert on the position and print out the wheel within the tests, next job is the schedule tests. Likely it is completely working with what I have, but I am still poking around to understand it.